File: Projects/cloned_repos/Date 2025-03-18 21s 28m 04s/kiss-slam/kiss_slam/occupancy_mapper.py
# MIT License
#
# Copyright (c) 2025 Tiziano Guadagnino, Benedikt Mersch, Saurabh Gupta, Cyrill
# Stachniss.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
import os

import numpy as np
import open3d as o3d
import yaml
from kiss_icp.voxelization import voxel_down_sample
from PIL import Image

from kiss_slam.config import OccupancyMapperConfig
from kiss_slam.kiss_slam_pybind import kiss_slam_pybind


class OccupancyGridMapper:
    def __init__(
        self,
        config: OccupancyMapperConfig,
    ):
        self.config = config
        self.occupancy_mapping_pipeline = kiss_slam_pybind._OccupancyMapper(
            self.config.resolution, self.config.max_range
        )

    def integrate_frame(self, frame: np.ndarray, pose: np.ndarray):
        frame_downsampled = voxel_down_sample(frame, self.config.resolution).astype(np.float32)
        self.occupancy_mapping_pipeline._integrate_frame(
            kiss_slam_pybind._Vector3fVector(frame_downsampled), pose
        )

    def compute_3d_occupancy_information(self):
        active_voxels, occupancies = self.occupancy_mapping_pipeline._get_active_voxels()
        self.active_voxels = np.asarray(active_voxels, int)
        self.occupancies = np.asarray(occupancies, float)
        self.occupied_voxels = self.active_voxels[
            np.where(self.occupancies > self.config.occupied_threshold)[0]
        ]

    def compute_2d_occupancy_information(self):
        min_z_idx = int(self.config.z_min // self.config.resolution)
        max_z_idx = int(self.config.z_max // self.config.resolution)

        indices_in_range = np.where(
            (self.active_voxels[:, 2] <= max_z_idx) & (self.active_voxels[:, 2] >= min_z_idx)
        )[0]
        voxels_in_range = self.active_voxels[indices_in_range]

        self.lower_bound = np.min(voxels_in_range, 0)
        self.upper_bound = np.max(voxels_in_range, 0)
        nrows, ncols, nslices = self.upper_bound - self.lower_bound + 1
        occupancy_grid = np.ones((nrows, ncols, nslices)) * 0.5

        occupancy_grid[
            voxels_in_range[:, 0] - self.lower_bound[0],
            voxels_in_range[:, 1] - self.lower_bound[1],
            voxels_in_range[:, 2] - self.lower_bound[2],
        ] = self.occupancies[indices_in_range]
        self.occupancy_grid = 1.0 - np.max(occupancy_grid, 2)

    def write_3d_occupancy_grid(self, output_dir):
        map_points = (0.5 + self.occupied_voxels) * self.config.resolution
        o3d_pcd = o3d.geometry.PointCloud(o3d.utility.Vector3dVector(map_points))
        o3d_pcd.estimate_normals()
        o3d.io.write_point_cloud(os.path.join(output_dir, "occupancy_pcd.ply"), o3d_pcd)
        self.occupancy_mapping_pipeline._save_occupancy_volume(
            os.path.join(output_dir, "occupancy_grid_bonxai.bin")
        )

    def write_2d_occupancy_grid(self, output_dir):
        image_filename = "map.png"
        occupancy_image = np.rot90(self.occupancy_grid, 0) * 255.0
        Image.fromarray(np.asarray(occupancy_image, np.uint8)).save(
            os.path.join(output_dir, image_filename)
        )
        grid_info = {
            "image": image_filename,
            "resolution": self.config.resolution,
            "origin": [
                float(self.lower_bound[0]) * self.config.resolution,
                float(self.lower_bound[1]) * self.config.resolution,
                0.0,
            ],
            "occupied_thresh": self.config.occupied_threshold,
            "free_thresh": self.config.free_threshold,
            "negate": 0,
        }
        with open(os.path.join(output_dir, "map.yaml"), "w") as yaml_file:
            yaml.dump(grid_info, yaml_file, default_flow_style=False, sort_keys=False)

File: Projects/cloned_repos/Date 2025-03-18 21s 28m 04s/kiss-slam/kiss_slam/kiss_slam_pybind/occupancy_mapper/occupancy_mapper.cpp
// MIT License

// Copyright (c) 2025 Tiziano Guadagnino, Benedikt Mersch, Saurabh Gupta,
// Meher Malladi Cyrill Stachniss.

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
#include "occupancy_mapper.hpp"

#include <Eigen/Core>
#include <algorithm>
#include <bonxai/bonxai.hpp>
#include <cmath>
#include <tuple>
#include <vector>

namespace {
static constexpr float logodds_free = -0.8473f;
static constexpr float logodds_occ = 2.1972f;
inline float ProbabilityOccupied(const float logodds) {
    return 1.0f - (1.0f / (1.0f + std::exp(logodds)));
}
}  // namespace

namespace occupancy_mapper {
OccupancyMapper::OccupancyMapper(const float resolution, const float max_range)
    : max_range_(max_range), map_(resolution), accessor_(map_.createAccessor()) {}

void OccupancyMapper::IntegrateFrame(const Vector3fVector &pointcloud,
                                     const Eigen::Matrix4f &pose) {
    const Eigen::Matrix3f &R = pose.block<3, 3>(0, 0);
    const Eigen::Vector3f &t = pose.block<3, 1>(0, 3);
    const auto start_coord = map_.posToCoord(t);
    std::for_each(pointcloud.cbegin(), pointcloud.cend(), [&](const Eigen::Vector3f &point) {
        const auto point_range = point.norm();
        if (point_range < max_range_) {
            const Eigen::Vector3f point_tf = R * point + t;
            const auto end_coord = map_.posToCoord(point_tf);
            Bresenham3DLine(start_coord, end_coord);
        }
    });
}

std::tuple<Vector3iVector, std::vector<float>> OccupancyMapper::GetOccupancyInformation() const {
    Vector3iVector voxel_indices;
    std::vector<float> voxel_occupancies;
    const auto num_of_active_voxels = map_.activeCellsCount();
    voxel_indices.reserve(num_of_active_voxels);
    voxel_occupancies.reserve(num_of_active_voxels);
    map_.forEachCell([&](const float &logodds, const Bonxai::CoordT &voxel) {
        voxel_indices.emplace_back(Bonxai::ConvertPoint<Eigen::Vector3i>(voxel));
        voxel_occupancies.emplace_back(ProbabilityOccupied(logodds));
    });
    return std::make_tuple(voxel_indices, voxel_occupancies);
}

void OccupancyMapper::UpdateVoxelOccupancy(const Bonxai::CoordT &coord, const float value) {
    // tg exploit Vdb caching
    accessor_.setCellOn(coord, 0.0);
    float *logodds = accessor_.value(coord);
    *logodds += value;
}

void OccupancyMapper::Bresenham3DLine(const Bonxai::CoordT &start_coord,
                                      const Bonxai::CoordT &end_coord) {
    const auto ray = end_coord - start_coord;
    const int x_sign = ray.x > 0 ? 1 : -1;
    const int y_sign = ray.y > 0 ? 1 : -1;
    const int z_sign = ray.z > 0 ? 1 : -1;

    const int ray_x_abs = x_sign * ray.x;
    const int ray_y_abs = y_sign * ray.y;
    const int ray_z_abs = z_sign * ray.z;

    int p1 = 0;
    int p2 = 0;
    Bonxai::CoordT delta{0, 0, 0};
    if (ray_x_abs >= ray_y_abs && ray_x_abs > ray_z_abs) {
        p1 = 2 * ray_y_abs - ray_x_abs;
        p2 = 2 * ray_z_abs - ray_x_abs;
        while (std::abs(delta.x) < ray_x_abs) {
            const auto voxel = start_coord + delta;
            UpdateVoxelOccupancy(voxel, logodds_free);
            if (p1 >= 0) {
                delta.y += y_sign;
                p1 -= (2 * ray_x_abs);
            }
            if (p2 >= 0) {
                delta.z += z_sign;
                p2 -= (2 * ray_x_abs);
            }
            p1 += (2 * ray_y_abs);
            p2 += (2 * ray_z_abs);
            delta.x += x_sign;
        }
    } else if (ray_y_abs >= ray_x_abs && ray_y_abs > ray_z_abs) {
        p1 = 2 * ray_x_abs - ray_y_abs;
        p2 = 2 * ray_z_abs - ray_y_abs;
        while (std::abs(delta.y) < ray_y_abs) {
            const auto voxel = start_coord + delta;
            UpdateVoxelOccupancy(voxel, logodds_free);
            if (p1 >= 0) {
                delta.x += x_sign;
                p1 -= (2 * ray_y_abs);
            }
            if (p2 >= 0) {
                delta.z += z_sign;
                p2 -= (2 * ray_y_abs);
            }
            p1 += (2 * ray_x_abs);
            p2 += (2 * ray_z_abs);
            delta.y += y_sign;
        }
    } else {
        p1 = 2 * ray_y_abs - ray_z_abs;
        p2 = 2 * ray_x_abs - ray_z_abs;
        while (std::abs(delta.z) < ray_z_abs) {
            const auto voxel = start_coord + delta;
            UpdateVoxelOccupancy(voxel, logodds_free);
            if (p1 >= 0) {
                delta.y += y_sign;
                p1 -= (2 * ray_z_abs);
            }
            if (p2 >= 0) {
                delta.x += x_sign;
                p2 -= (2 * ray_z_abs);
            }
            p1 += (2 * ray_y_abs);
            p2 += (2 * ray_x_abs);
            delta.z += z_sign;
        }
    }
    UpdateVoxelOccupancy(end_coord, logodds_occ);
}
}  // namespace occupancy_mapper
